src/main/java/uk/gov/hmcts/ccd/data/ReferenceDataRepository.java

```diff
@@ -0,0 +1,104 @@
+import java.util.List;
+import java.util.Optional;
+
+@Named
+@Slf4j
+public class ReferenceDataRepository {
+    private final SecurityUtils securityUtils;
+    private final RestTemplate restTemplate;
+    private final ApplicationParams applicationParams;
+    private final CacheManager cacheManager;
+
+    private static final String BUILDING_LOCATIONS_CACHE = "buildingLocations";
+    private static final String SERVICES_CACHE = "orgServices";
+    private static final String CACHE_KEY = "#root.method.name";
+
+    static final String BUILDING_LOCATIONS_PATH = "/refdata/location/building-locations";
+    static final String SERVICES_PATH = "/refdata/location/orgServices";
+
+    @Inject
+    public ReferenceDataRepository(final SecurityUtils securityUtils,
+                                   @Qualifier("restTemplate") final RestTemplate restTemplate,
+                                   final ApplicationParams applicationParams,
+                                   final CacheManager cacheManager) {
+        this.securityUtils = securityUtils;
+        this.restTemplate = restTemplate;
+        this.applicationParams = applicationParams;
+        this.cacheManager = cacheManager;
+    }
+
+    @Cacheable(cacheNames = BUILDING_LOCATIONS_CACHE, key = CACHE_KEY)
+    public List<BuildingLocation> getBuildingLocations() {
+        return getReferenceData(BUILDING_LOCATIONS_PATH, BuildingLocation[].class);
+    }
+
+    @Cacheable(cacheNames = SERVICES_CACHE, key = CACHE_KEY)
+    public List<Service> getServices() {
+        return getReferenceData(SERVICES_PATH, Service[].class);
+    }
+
+    private <T> List<T> getReferenceData(final String path, final Class<T[]> responseType) {
+        try {
+            final T[] result = restTemplate.exchange(
+                    applicationParams.getReferenceDataApiUrl() + path,
+                    HttpMethod.GET,
+                    new HttpEntity<>(securityUtils.authorizationHeaders()),
+                    responseType)
+                .getBody();
+
+            return Optional.ofNullable(result)
+                .map(Arrays::asList)
+                .orElse(Collections.emptyList());
+        } catch (RestClientException e) {
+            log.error("Error fetching reference data: ", e);
+            return Collections.emptyList();
+        }
+    }
+
+    @Scheduled(cron = "${reference.data.cache.refresh.rate.cron}")
+    public void updateBuildingLocationCache() {
+        final List<BuildingLocation> buildingLocations = getBuildingLocations();
+        updateCache(BUILDING_LOCATIONS_CACHE, buildingLocations);
+    }
+
+    @Scheduled(cron = "${reference.data.cache.refresh.rate.cron}")
+    public void updateServicesCache() {
+        final List<Service> services = getServices();
+        updateCache(SERVICES_CACHE, services);
+    }
+
+    private <T> void updateCache(final String cacheName, final List<T> newValue) {
+        if (!newValue.isEmpty()) {
+            clearCache(cacheName);
```

Comment:
do we need clearCache if put is overriding??

Reply:
Yep, you're right - put suffices

Reply:
Not clearing the cache before `put` seem to have some effect - some tests fail intermittently - perhaps due: 
`Actual registration may be performed in an asynchronous or deferred fashion, with subsequent lookups possibly not seeing the entry yet. This may for example be the case with transactional cache decorators. Use putIfAbsent for guaranteed immediate registration.`
I'm keeping the `clearCache(xxx)` and switching from `put` to `putIfAbsent`